docs/hugo/content/design/ADR-2023-04-Patch-Collections.md

```diff
@@ -0,0 +1,115 @@
+---
+
+## Context
+
+Some Azure Resource Providers do not quite have full replace semantics for a PUT.
+
+Instead, they have PATCH-like behavior on PUT which effectively boils down to: They will replace sub-objects/sub-collections if the object/collection is in the JSON payload, but they will leave untouched omitted properties or collections.
+
+To illustrate, assume you have a Person object with a Name and a list of Addresses.
+
+* If you PUT a Person object with a new Name and _no addresses_, the new name will be persisted, but the addresses will remain _unchanged_.
+* If you PUT a Person object with a new Name and an _empty list_ of addresses, the new name will be persisted, and the addresses collection will be cleared.
+
+This is a problem for ASO because we use `omitempty` on slice/map properties, which means that if you remove the last item from a slice/map, the property will be entirely omitted from the JSON payload.
+
+In principle, an ASO custom resource (CR) should be the goal-state for the resource and should be a complete definition of the desired state. Whether the collection is empty or missing entirely should not matter, we should interpret that as "clear the collection".
+
+### Option 1: Optional Properties
+
+Tack an extra pointer onto every slice/map property across all generated types.
+
+Instead of:
+
+``` go
+MyProperty []Type 'json:"myProperty,omitempty"'
+```
+
+we would generate:
+
+``` go
+MyProperty *[]Type 'json:"myProperty,omitempty"'
+```
+
+* PRO: Behaves exactly like we want from a serialization perspective.
+* CON: Horrible to use in code because you must dereference the property to range over it (or really do much of anything with it).
+* CON: Breaking change for any consumers of our object model.
+* CON: Modifies our Spec and Status object definitions for an issue that is specific to ARM
+
+### Option 2: Remove `omitempty` 
+
+Remove omitempty from slice/map properties across all generated types.
+
+``` go
+MyProperty []Type 'json:"myProperty"'. 
+```
+
+This means that the slice or map will always be serialized, even if empty.
+
+* PRO: Behaves exactly like we want from a serialization perspective.
+* PRO: Not a breaking change for consumers of our object model.
+* CON: Empty collections will always be present when an ASO CR is read from the cluster, even if the user did not specify them at all.
+* CON: If we extend this to all properties, all properties will be present when an ASO CR is read from the cluster, even if the user did not specify them at all.
+* CON: Modifies our Spec and Status object definitions for an issue that is specific to ARM
+* QUERY: It is possible that some Azure services treat `myProperty: null` as different than omitting the property entirely, although we do not actually know of any cases where that is true.
+
+### Option 3: Custom JSON Serialization
+
+Use a custom JSON serialization library that supports an `omitnil` equivalent, such as [jettison](https://github.com/wI2L/jettison).
```

Comment:
Does this solve non-collection properties issue? I'm not sure that it does. The only way to clear those properties may be to explicitly pass json `null` to the server for the field. `omitnil` would prevent that. `omitnil` saves us for collections because we can still pass the empty collection.

Reply:
I've added that as an outstanding question.